Simple benchmarks

TotalVerb opened the pull request:
It would be helpful to be able to more precisely measure the impact of particular optimizations, so I've imported some benchmarks from the MIT-licensed [nativejson-benchmark](https://github.com/miloyip/nativejson-benchmark). This is very rough, but it's a starting point.

I also added some compliance tests. We're not 100% compliant, but these tests will help pinpoint possible regressions, and also 100% compliance is something to work towards.

kmsquire commented:
Haven't looked through thoroughly, but at a glance, 👍 !

TotalVerb commented:
Since they were there anyway, I added the roundtrips to the tests.

tkelman commented:
how about downloading a tarball instead of inflating the repo size so dramatically?

TotalVerb commented:
@tkelman Sure. Will work on that later today.

AppVeyor on 32-bit is apparently failing because of an overflow. That's not really desirable in any case. I think a reasonable patch in the interim would be to use `Int64` always, since that supports all JSON numbers, and fix the numeric parsing later.

TotalVerb commented:
On second thought, this problem was discovered by a test but not introduced by this PR, so skipping the test seems more reasonable. It will be fixed if number parsing becomes type-stable.

tkelman commented on test/json-checker.jl:
```diff
@@ -0,0 +1,35 @@
+# Run JSON checker tests, excluding tests fail01 and fail18
+
+const JSON_DATA_DIR = joinpath(Pkg.dir("JSON"), "data")
```
don't use Pkg.dir, the package may be installed and used from somewhere else. `dirname(@__FILE__)` is better

TotalVerb commented:
Thanks for the comments. I've made it download the benchmark files when/if required. I also added a citylots file which is much larger.

TotalVerb commented:
Last try didn't fix the problem. This try should.

TotalVerb commented:
Well, that still didn't work. I'll have another look tomorrow.

TotalVerb commented:
Finally passes Appveyor.

tkelman commented on test/runtests.jl:
```diff
@@ -19,11 +20,12 @@ validate_c(c) = begin
                 end
 
 validate_e(e) = begin
+                    T = TypeVar(:T, AbstractString)
```
odd, why's this necessary?

TotalVerb replied:
I don't think it is. That's an unrelated change that accidentally made it in. I'll take it out.

TotalVerb commented:
Looks like all tests are passing now. Anything else to handle?

tkelman commented on test/json-checker.jl:
```diff
@@ -0,0 +1,41 @@
+
+        # should not reach here
+        print_with_color(:red, " [Noncompliant] ")
+        println("Expecting $file to fail, but parsed without error")
```
should this cause an error?

TotalVerb replied:
Those tests don't work yet though. I would use @test_broken if that were available, but it hasn't landed yet.

tkelman replied:
ah okay. add a #FIXME comment saying to make it an error in the future?

tkelman commented:
This LGTM aside from the FIXME thing. Will see if anyone else has any opinions, if not ping again if I haven't merged tomorrow.

TotalVerb commented:
Added two FIXME comments and some better error testing. Also added two extra tests to make the numbers nice (previously we were skipping 1 and 18).

A lot of these will be fixed by the parser optimization I'm working on, so hopefully we'll be able to remove the excess code soon enough.

tkelman commented on test/json-checker.jl:
```diff
@@ -0,0 +1,62 @@
+
+            print_with_color(:red, " [Fail] ")
+            println("Test $file parsed successfully; should throw ErrorException.")
+            println(e)
```
what's `e` ?

TotalVerb replied:
Oops. Nice catch.

tkelman commented on test/json-checker.jl:
```diff
@@ -0,0 +1,62 @@
+            println("Test $file parsed successfully; should throw ErrorException.")
+            println(e)
+        catch e
+            if isa(e, ErrorException)
```
Doesn't need to be fixed right now, but I wonder if the exception types thrown for various invalid data or parsing failures should be made more specific.

TotalVerb commented:
These 32-bit issues are kind of annoying to test. I hope this squashes it?

tkelman commented:
It looks like. What OS do you use?

TotalVerb commented:
I'm on Linux, but I have access to a Windows box. Both are 64-bit.

tkelman commented:
You should be able to download the binaries for win32 and linux32 and run them on 64 bit systems. Packages can get confused if there are binary dependencies, but that's not an issue here.

nalimilan commented:
> how about downloading a tarball instead of inflating the repo size so dramatically?

Does this really make a large difference in cloning time? The files are very small, and less than 300 lines in total.

tkelman commented:
An earlier version of this PR was adding over 60000 lines. It's fine now.

tkelman commented:
Thank you for the contribution!
